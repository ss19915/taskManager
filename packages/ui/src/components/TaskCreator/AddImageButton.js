import React from 'react';
import { Fab, AddAPhotoIcon } from '@task-manager/theme';
import T from 'prop-types';

const AddImageButton = (props) => {
    const { onChange } = props;
    const inputRef = React.createRef();

    const openFileChooser = () => { inputRef.current.click() };
    const getObjectValues = (object) => Object.values(object);
    return (
        <React.Fragment>
            <input
                hidden multiple
                ref={inputRef}
                type='file'
                onChange={({ target }) => onChange(getObjectValues(target.files))}
            />
            <Fab onClick={openFileChooser}>
                <AddAPhotoIcon />
            </Fab>
        </React.Fragment>
    );
};

AddImageButton.propTypes = {
    onChange: T.func.isRequired,
};

export default AddImageButton;