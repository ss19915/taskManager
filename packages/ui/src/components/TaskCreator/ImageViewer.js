import React from 'react';
import { Card, IconButton, CloseIcon } from '@task-manager/theme';
import T from 'prop-types';
import utils from '../../utils'

const getImageFromPath = (path) => URL.createObjectURL(path);

const ImageViewer = (props) => {
    const { imageFile, removeImage, name } = props;
    const maxTitleLenght = 10;

    return (
        <React.Fragment>
        <Card>
            <Card.Header
                    action={
                        <IconButton onClick={() => removeImage(name)}>
                            <CloseIcon />
                        </IconButton>
                    }
                    title={utils.limitCharacter(imageFile.name, maxTitleLenght)}
                />
            <Card.ActionArea>
                <Card.Media>
                    <img width='100%' src={getImageFromPath(imageFile)} />
                </Card.Media>
            </Card.ActionArea>
        </Card>
        </React.Fragment>
    );
};

ImageViewer.propTypes = {
    imageFile: T.object.isRequired,
    name: T.any.isRequired,
    removeImage: T.func.isRequired,
};


export default ImageViewer;