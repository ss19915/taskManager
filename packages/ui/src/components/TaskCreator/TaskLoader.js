import React from 'react';
import { Modal, Card, Typography, CircularProgress, Button, Grid } from '@task-manager/theme';
import { API_STATUS } from '../../constants';

const TaskLoader = (props) => {
    const { loading, error, createNewTask, retry, redirectHome, cancelCreateTask } = props;

    const Loader = () => (
        <Card.Content>
            <Typography display='inline'>Saving</Typography>
            <CircularProgress />
        </Card.Content>
    );

    const SaveMenu = () => (
        <React.Fragment>
            <Card.Content>
                <Typography>  Task Created Successfully</Typography>
            </Card.Content>
            <Card.Actions>
                <Button onClick={createNewTask}>Create More</Button>
                <Button onClick={redirectHome}>Home</Button>
            </Card.Actions>
        </React.Fragment>
    );

    const ErrorMenu = () => (
        <React.Fragment>
            <Card.Content>
                <Typography>{JSON.stringify(error)}</Typography>
            </Card.Content>
            <Card.Actions>
                <Button onClick={retry}>Retry</Button>
                <Button onClick={cancelCreateTask}>Cancel</Button>
            </Card.Actions>
        </React.Fragment>
    );
    return (
        <Modal open={loading !== API_STATUS.INITIAL}>
            <Grid
                container
                justify='center'
                alignItems='center'
            >
                <Grid item>
                    <Card>
                        {loading === API_STATUS.SENT && <Loader />}
                        {loading === API_STATUS.RECEIVED && <SaveMenu />}
                        {loading === API_STATUS.ERROR && <ErrorMenu />}
                    </Card>
                </Grid>
            </Grid>
        </Modal>
    );
}

export default TaskLoader;