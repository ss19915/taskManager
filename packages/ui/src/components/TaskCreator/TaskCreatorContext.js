import React from 'react';
import TaskCreator from './TaskCreator';
import { withRouter } from 'react-router-dom';
import constants from '../../constants';

const TaskCreatorContext = (props) => {
    const { history } = props;
    const redirectHome = () => {
        history.push(constants.ROUTES.DASHBOARD);
    };

    return (
        <TaskCreator
            {...props}
            redirectHome={redirectHome}
        />
    );
}


export default withRouter(TaskCreatorContext);