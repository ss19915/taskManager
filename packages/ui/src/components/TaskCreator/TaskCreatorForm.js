import React from 'react';
import { Card, TextField, Fab, Typography, Grid, ArrowBackIcon } from '@task-manager/theme';
import T from 'prop-types';
import AddImageButton from './AddImageButton';
import ImageViewer from './ImageViewer';

const TaskCreatorForm = (props) => {
    const {
        createTask,
        onImageAdd,
        onNameChange,
        onDescriptionChange,
        imageFiles,
        removeImage,
        isCreateDisabled,
        goBack
    } = props;

    const MultipleImageViewer = () => (
        <Grid spacing={1} container>
            {imageFiles.map((imageFile, index) => (
                <Grid item xl={1} lg={2} md={2} sm={3} xs={6} key={index}>
                    <ImageViewer
                        imageFile={imageFile}
                        name={index}
                        removeImage={removeImage}
                    />
                </Grid>
            ))}
        </Grid>
    );

    return (
        <Card>
            <Card.Header
                title='Task Creator'
                avatar={
                    <Fab onClick={goBack}><ArrowBackIcon/></Fab>
                }
                titleTypographyProps={{
                    variant: 'h5'
                }}
                action={
                    <Fab
                        color='primary'
                        variant='extended'
                        onClick={createTask}
                        disabled={isCreateDisabled}
                    >
                        Create Task
                    </Fab>
                }
            />
            <Card.Content>
                <Typography>Name*</Typography>
                <TextField
                    fullWidth
                    onChange={({ target: { value } }) => onNameChange(value)}
                />
                <Typography>Description</Typography>
                <TextField fullWidth onChange={({ target: { value } }) => onDescriptionChange(value)} multiline />
                <MultipleImageViewer />
            </Card.Content>
            <Card.Actions>
                <AddImageButton onChange={onImageAdd} />
            </Card.Actions>
        </Card>
    );
};

TaskCreatorForm.propTypes = {
    createTask: T.func.isRequired,
    imageFiles: T.arrayOf(T.object),
    isCreateDisabled: T.bool.isRequired,
    onDescriptionChange: T.func.isRequired,
    onImageAdd: T.func.isRequired,
    onNameChange: T.func.isRequired,
    removeImage: T.func.isRequired,
};

export default TaskCreatorForm