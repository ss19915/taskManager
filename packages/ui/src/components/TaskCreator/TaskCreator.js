import React from 'react';
import TaskCreatorForm from './TaskCreatorForm';
import _ from 'lodash';

class TaskCreator extends React.PureComponent {
    state = {
        imageFiles: [],
        isCreateDisabled: true,
    };

    createTask = () => { };

    onDescriptionChange = (taskDescription) => this.setState({ taskDescription });

    removeImage = (index) => {
        const imageFiles = _.cloneDeep(this.state.imageFiles);

        imageFiles.splice(index, 1);
        this.setState({ imageFiles });
    };

    removeInvalidImageFiles = (imageFiles) => {
        const validImageFiles = [];

        imageFiles.map((imageFile) => {
            if (imageFile.type === 'image/jpeg') {
                validImageFiles.push(imageFile);
            }
        });
        return (validImageFiles);
    };

    onImageAdd = (filePaths) => {
        const imageFiles = this.removeInvalidImageFiles(filePaths);;

        this.setState((prevState) => ({
            imageFiles: [...prevState.imageFiles, ...imageFiles],
        }));
    };

    onNameChange = (taskName) => {
        this.setState( (prevState) => {
            const newState = { taskName };

            if(prevState.isCreateDisabled && !_.isEmpty(taskName)){
                newState.isCreateDisabled = false;
            }
            else if( !prevState.isCreateDisabled && _.isEmpty(taskName)){
                newState.isCreateDisabled = true;
            }

            return(newState);
        })
    };

    render() {
        const { imageFiles, isCreateDisabled } = this.state;

        return (
            <TaskCreatorForm
                onDescriptionChange={this.onDescriptionChange}
                onImageAdd={this.onImageAdd}
                onNameChange={this.onNameChange}
                createTask={this.createTask}
                imageFiles={imageFiles}
                removeImage={this.removeImage}
                isCreateDisabled={isCreateDisabled}
            />
        );
    }
}

export default TaskCreator;