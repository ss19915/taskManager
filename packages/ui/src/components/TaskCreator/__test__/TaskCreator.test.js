import TaskCreator from '../TaskCreator';
import React from 'react';

describe('Component: TaskCreator', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <TaskCreator
                saveTask={() => { }}
                redirectHome={() => {}}
            />
        );

    });

    it('Should match snapshot', () => {

        expect(wrapper).toMatchSnapshot();
    });

    it('should remove image on remoceImage call', () => {
        const imageFiles = [
            { type: 'image/jpeg', name: '0.jpg' },
            { type: 'image/jpeg', name: '1.jpg' },
            { type: 'image/jpeg', name: '2.jpg' },
            { type: 'image/jpeg', name: '3.jpg' },
            { type: 'image/jpeg', name: '4.jpg' },
            { type: 'image/jpeg', name: '5.jpg' },
        ];
        const removeImageIndex = 2;
        const instance = wrapper.instance();
        instance.onImageAdd(imageFiles);
        instance.removeImage(removeImageIndex);

        imageFiles.splice(removeImageIndex, 1);

        expect(instance.state.imageFiles).toEqual(imageFiles);
    });

    it('should remove Invalid image files', () => {
        const imageFiles = [
            { type: 'image/jpeg', name: '0.jpg' },
            { type: 'image/jpeg', name: '1.jpg' },
            { type: 'image/jpeg', name: '2.jpg' },
            { type: 'pdf', name: '3.pdf' },
            { type: 'image/jpeg', name: '4.jpg' },
            { type: 'image/jpeg', name: '5.jpg' },
        ];
        const invalidImageIndex = 3;
        const instance = wrapper.instance();

        instance.onImageAdd(imageFiles);
        imageFiles.splice(invalidImageIndex, 1);
        expect(instance.state.imageFiles).toEqual(imageFiles);
    });
});