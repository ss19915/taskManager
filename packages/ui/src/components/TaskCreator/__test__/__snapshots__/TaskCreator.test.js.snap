// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TaskCreator Should match snapshot 1`] = `
<TaskCreatorForm
  createTask={[Function]}
  imageFiles={Array []}
  isCreateDisabled={true}
  onDescriptionChange={[Function]}
  onImageAdd={[Function]}
  onNameChange={[Function]}
  removeImage={[Function]}
/>
`;
